src/slice.c

```diff
@@ -381,11 +375,15 @@ bool vec_is_restored(r_obj* x, r_obj* to) {
     return false;
   }
 
-  // Class is restored if it contains any other attributes than names.
-  // We might want to add support for data frames later on.
+  // Class is restored if it contains any other attributes than names, dim, or
+  // dimnames. We might want to add support for data frames later on.
   r_obj* node = attrib;
   while (node != r_null) {
-    if (r_node_tag(node) == r_syms.names) {
+    r_obj* tag = r_node_tag(node);
+
+    if (tag == r_syms.names ||
+        tag == r_syms.dim ||
+        tag == r_syms.dim_names) {
```

Comment:
`vec_is_restored()` had this bug where it said "you don't need to restore" if the object had `dim` or `dimname` attributes, which definitely isn't true. Those are two of the attributes that base R `[` methods will propagate, like names (but unlike the class).